corehq/motech/value_source.py

```diff
@@ -377,36 +418,22 @@ def _get_commcare_value(self, case_trigger_info):
             return get_ancestor_location_metadata_value(location, self.location_field)
 
 
-class JsonPathMixin(DocumentSchema):
-    """
-    Used for importing a value from a JSON document.
-    """
-    jsonpath = StringProperty(required=True, validators=not_blank)
-
-    def _get_external_value(self, external_data):
-        jsonpath = parse_jsonpath(self.jsonpath)
-        matches = jsonpath.find(external_data)
-        values = [m.value for m in matches]
-        if not values:
-            return None
-        elif len(values) == 1:
-            return values[0]
-        else:
-            return values
-
-    def get_import_value(self, external_data):
-        external_value = self._get_external_value(external_data)
-        return self.deserialize(external_value)
-
-
 class JsonPathCaseProperty(CaseProperty, JsonPathMixin):
-    pass
+
+    def __init__(self, *args, **kwargs):
+        DocumentSchema.__init__(self, *args, **kwargs)
+        warn("JsonPathCaseProperty is deprecated; use CaseProperty", DeprecationWarning)
 
 
 class JsonPathCasePropertyMap(CasePropertyMap, JsonPathMixin):
-    pass
+
+    def __init__(self, *args, **kwargs):
+        DocumentSchema.__init__(self, *args, **kwargs)
+        warn("JsonPathCasePropertyMap is deprecated; use CaseProperty", DeprecationWarning)
 
 
+# This class is to be replaced by CasePropertyConstantValue.
+# cf. nh/omrs/case_const (PR: https://github.com/dimagi/commcare-hq/pull/25925)
 class JsonPathConstantValue(ConstantValue, JsonPathMixin):
 
     def _get_external_value(self, external_data):
```

Comment:
It was not a great class.

(I would have said "protected", because private methods have a [dunder prefix](https://radek.io/2011/07/21/private-protected-and-public-in-python/). As a protected method, isn't it considered OK to override in subclasses?)

Reply:
I hadn't heard of `_` (single underscore) indicating "protected" status before I read that article. My understanding is that any method beginning with an underscore is "private" (by convention only, of course, there is no real privacy in Python). Then again we are consenting adults, and sometimes you need to bend the rules. Practicality beats purity.

As for dunder-prefix name-mangling privacy/magic, don't use magic unless you really need it (and you almost never do). See also this [great SO answer on privacy in Python](https://stackoverflow.com/a/7456865/10840).

Reply:
I agree with a lot of what that SO answer says. 

I don't think of accessing dunder-prefixed properties and methods as magic. I think the dunder prefix indicates intention, and name mangling is to make it clear that code that uses it is violating that intention. "Access without consent" in a sense, so ugly by design.

But [the answer which that answer links to](https://stackoverflow.com/questions/70528/why-are-pythons-private-methods-not-actually-private/70900#70900) is very persuasive, "The name scrambling is used to ensure that subclasses don't accidentally override the private methods and attributes of their superclasses. It's not designed to prevent deliberate access from outside."

I also wonder whether those SO answers suggest that intentionally overloading single-underscore-prefixed methods is generally OK?

I guess the real answer, for us (HQ devs) specifically, is whether we expect it. If we don't expect it, we should either not do it, or agree that it's OK, and so expect it, even if only now and then (because, like that answer suggests, default to public) in the future.

Reply:
> I don't think of accessing dunder-prefixed properties and methods as magic.

I don't think I said quite that, but I probably wasn't very clear about what I was trying to say. What I intended to say was _name mangling is magic_, and I avoid using that feature if possible. See the [Paste Style Guide](https://pythonpaste.readthedocs.io/en/latest/StyleGuide.html), which says:

> Never, ever use two leading underscores. This is annoyingly private. If name clashes are a concern, use explicit name mangling instead (e.g., _MyThing_blahblah). This is essentially the same thing as double-underscore, only it’s transparent where double underscore obscures.

Also, I hope we're not getting confused about built-in dunder methods like `__init__` and `__str__`, which are a very different thing from `x.__my_private_thing`; I was only talking about the latter.

> intentionally overloading single-underscore-prefixed methods is generally OK

Not sure what you meant by "intentionally overloading". Did you mean "intentionally overriding" as in creating a subclass that overrides a method whose name begins with a single underscore? ([disambiguation](https://www.programcreek.com/2009/02/overriding-and-overloading-in-java-with-examples/))

My feeling is that single-underscore-prefixed methods are private. Creating a subclass that overrides a private method tightly couples that subclass to the private implementation details of the base class, which makes it harder to maintain the base class. Tight coupling should be avoided.

Reply:
I think we agree on every single point but one; that properties and methods prefixed with a single underscore are private, not protected.

But just as English is defined by what is spoken not by what appears in a dictionary, so Python convention, in the context of the HQ codebase, is whatever HQ devs agree it is.

If single-underscore-prefixed properties and methods are considered private, then they should not be overridden (sorry) by subclasses.